ballerina-shell/modules/shell-cli/src/main/java/io/ballerina/shell/cli/jline/parser/ParserStateMachine.java

```diff
@@ -0,0 +1,317 @@
+/*
+ * Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.ballerina.shell.cli.jline.parser;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.Stack;
+
+/**
+ * State machine implementation for ballerina parser.
+ * Specifically used by {@link io.ballerina.shell.cli.jline.JlineBallerinaParser}.
+ * <p>
+ * In this stage, the "counting parentheses" approach is extended
+ * so that it is consistent with Ballerina syntax, and correct in the sense that if it
+ * says the input is incomplete, then in Ballerina it will definitely be incomplete.
+ * <p>
+ * There are several cases where it isn't at the moment e.g. {@code "\"(", // (}
```

Comment:
Is this still to be done? Do we need to create an issue?

Reply:
I think #28315 can track this. This PR is only a partial fix to that issue, solving it completely would require a parser API.